src/controller/CHIPDeviceController.cpp

```diff
@@ -686,9 +688,30 @@ ControllerDeviceInitParams DeviceController::GetControllerDeviceInitParams()
 
 DeviceCommissioner::DeviceCommissioner()
 {
-    mPairingDelegate      = nullptr;
-    mDeviceBeingPaired    = kNumMaxActiveDevices;
-    mPairedDevicesUpdated = false;
+    mPairingDelegate        = nullptr;
+    mDeviceBeingPaired      = kNumMaxActiveDevices;
+    mPairedDevicesUpdated   = false;
+    mOpCSRResponseCallback  = nullptr;
+    mOpCertResponseCallback = nullptr;
+    mOnFailureCallback      = nullptr;
+}
+
+DeviceCommissioner::~DeviceCommissioner()
+{
+    if (mOpCSRResponseCallback != nullptr)
```

Comment:
Why are these callbacks dynamically allocated?  Wouldn't it be simpler to have these as value fields of the `DeviceCommissioner` object?  It does not seem like a `DeviceCommissioner` object without these will be very useful.  Also, this pattern forces long-term residency for the callbacks in the heap.

Also, is there a lifecycle problem for the callbacks?  How can we be sure that no operation is in flight when the callback destructors are called?

Reply:
@msandstedt, I've updated the code to add callbacks to commissioner class as member variables, instead of ptrs.